### PR TITLE
Support MIMO GTP checkpoint resume

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## Summary
- use the GTP-inclusive data group for ordinary replicated MIMO checkpoint state
- include TP x GTP, PP, and DP x CP coordinates in MIMO optimizer replica IDs
- detect a MIMO `HybridModel` language spec on encoder-only ranks
- make the existing heterogeneous 20L launcher verify an exact normal save/load round-trip
- run the round-trip as a dedicated 1x8-GPU H100 functional test

## Base
One MIMO-specific commit on `main` at `d98e8a64dd`, including #6775.

## Regression
The existing 20L launcher performs a normal training save at iteration 1. A second invocation takes the normal checkpoint load path and uses a test-only hook to immediately resave the loaded state before another training step.

The comparator requires identical checkpoint keys and exact logical state for both model roles, both optimizer roles, optimizer parameters and moments, language/encoder RNG, scheduler/LR state, iteration, checkpoint version, FLOPs, and remaining common metadata.

Distributed-optimizer ranges explicitly marked `padding=True` are the only excluded bytes; every non-padding byte must match. This avoids comparing uninitialized `torch.empty` padding while retaining the default DP-reshardable checkpoint path.

No checkpoint conversion, `training.py`, dataloader change, or new shell wrapper is included.

## Validation
- CMH job `3239303`: 2 nodes x 4 GPUs, completed `0:0`
- exact normal save/load/resave round-trip: 752 entries / 36,663,494,528 bytes matched
- 36,457 differing bytes were confined to manifest-proven optimizer padding
- functional recipe uses one plain pytest process so the existing launcher owns its 8-rank phases
- the generic unit-test and functional-test harnesses are unchanged
- scoped Black, isort, Pylint, Ruff, `py_compile`, shell syntax, and diff check pass

### CODE REVIEW

File: megatron/training/checkpointing.py [hunk 1]
```diff
@@ -2386,6 +2387,10 @@ def _contains_hybrid_model(module):
         while module is not None:
             if isinstance(module, HybridModel):
                 return True
+            if isinstance(module, MimoModel):
```
Comment (yashaswikarnati) [inline megatron/training/checkpointing.py:2390]: we need this check here as current mamba/hyrbid model conversion fails on encoder ranks. encoder is neither mamba nor hybrid

**Review by deepakn94 — APPROVED** [review]

**Review by chtruong814 — APPROVED** [review]

File: tests/test_utils/recipes/h100/mimo-checkpoint.yaml
Comment (Phlip79) [inline tests/test_utils/recipes/h100/mimo-checkpoint.yaml:1]: Why not use existing `mimo.yaml`?

**Review by Phlip79 — APPROVED** [review]

File: tests/test_utils/recipes/h100/mimo-checkpoint.yaml
Comment (yashaswikarnati) [reply tests/test_utils/recipes/h100/mimo-checkpoint.yaml:1]: makes sense, will update it..

**Comment by yashaswikarnati** [comment]
/ok to test [486e0d9](https://github.com/NVIDIA/Megatron-LM/pull/6743/commits/486e0d99489227c9da58b7ef0629851592c8019b)

**Review by fanshiqing — APPROVED** [review]

**Review by asolergi-nv — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [76796dc](https://github.com/NVIDIA/Megatron-LM/pull/6743/commits/76796dc79782dda192c8db5ee007ce41571d5480)

**Comment by yashaswikarnati** [comment]
/ok to test [1ac7ab1](https://github.com/NVIDIA/Megatron-LM/pull/6743/commits/1ac7ab166c5a402cfad22c91d945d248d5fe7315)

**Review by maanug-nv — APPROVED** [review]
